storage/src/nodestore.rs

```diff
@@ -1428,6 +1430,73 @@ impl<S: WritableStorage> NodeStore<Committed, S> {
     }
 }
 
+pub(crate) const STORAGE_AREA_START: LinearAddress =
+    LinearAddress::new(NodeStoreHeader::SIZE).unwrap();
+
+/// NodeStore checker
+// TODO: S needs to be writeable if we ask checker to fix the issues
+impl<S: ReadableStorage> NodeStore<Committed, S> {
+    /// Go through the filebacked storage and check for any inconsistencies. It proceeds in the following steps:
+    /// 1. Check the header
+    /// 2. traverse the trie and check the nodes
+    /// 3. check the free list
+    /// 4. check missed areas - what are the spaces between trie nodes and free lists we have traversed?
+    // TODO: add merkle hash checks as well
+    pub async fn check(&self) -> Result<(), CheckerError> {
+        // 1. Check the header
+        let db_size = self.header.size;
+        let file_size = self.storage.size()?;
+        if db_size < file_size {
+            return Err(CheckerError::InvalidDBSize {
+                db_size,
+                description: format!(
+                    "db size should not be smaller than the file size ({file_size})"
+                ),
+            });
+        }
+
+        let mut visited = LinearAddressRangeSet::new(db_size)?;
+
+        // 2. traverse the trie and check the nodes
+        if let Some(root_address) = self.header.root_address {
+            // the database is not empty, traverse the trie
+            self.traverse_trie(root_address, &mut visited).await?;
+        };
+
+        // 3. check the free list - this can happen in parallel with the trie traversal
+
+        // 4. check missed areas - what are the spaces between trie nodes and free lists we have traversed?
+        let _ = visited.complement(); // TODO
+
+        Ok(())
+    }
+
+    /// Recursively traverse the trie from the given root address.
+    async fn traverse_trie(
+        &self,
+        subtree_root_address: LinearAddress,
+        visited: &mut LinearAddressRangeSet,
+    ) -> Result<(), CheckerError> {
+        let (_, area_size) = self.area_index_and_size(subtree_root_address)?;
+        visited.insert_area(subtree_root_address, area_size)?;
+
+        if let Node::Branch(branch) = self.read_node(subtree_root_address)?.as_ref() {
+            // this is an internal node, traverse the children
+            let child_addrs = branch.children.iter().filter_map(|child| match child {
+                None => None,
+                Some(Child::AddressWithHash(address, _)) => Some(*address),
+                _ => panic!("the child is not persisted yet, this should not happen"),
+            });
+
+            for child_addr in child_addrs {
```

Comment:
```suggestion
            for child_addr in branch.children_iter() {
```

Reply:
`children_iter()` gives a iterator over the idx and the hash of children existing in the node. To get the address I need to go back to the list of children with the returned idx anyways. Let me rearrange the code in another way...

Reply:
Ah, right, you need something different. Two options:

1. Create a two iterators for BranchNode: `children_hashes` and `children_addresses`, changing existing callers of `children_iter` to `children_hashes`.
2. Change `children_iter` to be a bit more general and return `impl Iterator<Item = (usize, (LinearAddress, &HashType))>` and change the 3 existing callers to do `children_iter().map(|(offset, (addr, hash))| (offset, hash))`